core/src/main/java/org/mule/runtime/core/internal/exception/TemplateOnErrorHandler.java

```diff
@@ -64,24 +63,64 @@
   @Override
   final public Event handleException(MessagingException exception, Event event) {
     try {
-      return new ExceptionMessageProcessor(exception, muleContext, flowConstruct).process(event);
-    } catch (MuleException e) {
-      throw new RuntimeException(e);
+      return from(applyInternal(exception, event)).block();
+    } catch (Throwable throwable) {
+      throw new RuntimeException(unwrap(throwable));
     }
   }
 
   @Override
   public Publisher<Event> apply(final MessagingException exception) {
-    Publisher<Event> publisher =
-        new ExceptionMessageProcessor(exception, muleContext, flowConstruct).apply(just(exception.getEvent()));
-    return from(publisher).handle((event, sink) -> {
-      if (exception.handled()) {
-        sink.next(event);
-      } else {
-        exception.setProcessedEvent(event);
-        sink.error(exception);
-      }
-    });
+    return applyInternal(exception, exception.getEvent());
+  }
+
+  private Publisher<Event> applyInternal(final MessagingException exception, Event event) {
+    return just(event)
+        .doOnNext(request -> {
```

Comment:
can some of these blocks be refactored to different methods for better readability?